### ISSUE TITLE
MetricsQL TimeSeries "or" operator not work like expected when metrics contains empty values

### ISSUE THREAD

**Issue by Nick135** [VictoriaMetrics/VictoriaMetrics#3710]
### Describe the bug

When I divide the result of a query, each line has the same value. Line looks like a combination of the values divided by constant.

### To Reproduce
```
WITH (
    data = {__name__=~"MyMeasurement_[A-Z].*",unit="°C"} > 0 < 120,
    data1 = {__name__=~"MyMeasurement_[A-Z].*",unit="°C"} > 120 < 1000
)
(data or (data1/10))
```

### Version

victoria-metrics-20230118-200413-tags-v1.86.2-0-g9785b3f57

Grafana v9.0.4 with Prometheus plugin

### Screenshots

**Original Data**
![image](https://user-images.githubusercontent.com/14072733/214550310-3beab738-2a8e-4dc8-af53-b8ba07797f0f.png)

**Divided by 10**
![image](https://user-images.githubusercontent.com/14072733/214550436-763b538d-b527-4d3c-8770-9e86cd560f36.png)

**Comment by Nick135** [comment]
Is there anything related to this that I can do to speed it up?

**Comment by valyala** [comment]
The query works as expected - see [this playground example](https://play.victoriametrics.com/select/accounting/1/6a716b0f-38bc-4856-90ce-448fd713e3fe/prometheus/graph/?g0.expr=with%20(q%3D30*running_sum(rand(1)-0.5)%2B200)%0A(%0A%20%20alias(q%2C%20%22original%22)%2C%0A%20%20alias((q%20%3C%20120)%20or%20(q%2F10)%2C%20%22calculated%22)%2C%0A)&g0.range_input=30m&g0.end_input=2023-01-28T01%3A14%3A05&g0.relative_time=last_30_minutes).

It looks like your original data contains unexpected raw samples, which lead to unexpected calculations. I'd suggest exporting raw samples from VictoriaMetrics and verifying whether they contain the expected values. See [how to export raw samples from VictoriaMetrics](https://valyala.medium.com/analyzing-prometheus-data-with-external-tools-5f3e5e147639). See also [troubleshooting docs](https://docs.victoriametrics.com/Troubleshooting.html#unexpected-query-results).

**Comment by Nick135** [comment]
I test this example from the docs: [How to export data in JSON line format](https://github.com/VictoriaMetrics/VictoriaMetrics/wiki/Single-server-VictoriaMetrics#how-to-export-data-in-json-line-format)

But I get 'missing `match[]` arg'
What do I wrong to export the raw data?
**Command**
```
curl http://<victoriametrics-addr>:8428/api/v1/export -d 'match[]={__name__!=""}' -d 'start=2022-06-06T19:25:48+00:00' -d 'end=2022-06-06T19:29:07+00:00'
```
**Result**
```
remoteAddr: "192.168.214.198:62077"; requestURI: /api/v1/export?%27end=2022-06-06T19%3A29%3A07+00%3A00%27&%27match%5B%5D=%7B__name__%21%3D%7D%27&%27start=2022-06-06T19%3A25%3A48+00%3A00%27; missing `match[]` arg
```

**Comment by Nick135** [comment]
This is the raw data result from Grafana:
### {__name__=~"SN003:1_[A-Z].*",unit="°C"}
[Test-data-as-seriestocolumns.csv](https://github.com/VictoriaMetrics/VictoriaMetrics/files/10538780/Test-data-as-seriestocolumns.csv)
![image](https://user-images.githubusercontent.com/14072733/215563625-96c5bea4-028c-49bb-9d75-83604a47fad3.png)

### WITH (data or data1)
[Test-data-as-seriestocolumns-WITH-or.csv](https://github.com/VictoriaMetrics/VictoriaMetrics/files/10538840/Test-data-as-seriestocolumns-WITH-or.csv)

### Compare
![image](https://user-images.githubusercontent.com/14072733/215564773-3fa0398c-3925-431f-8c4d-e1aa5c12ba83.png)

**The Result with `or` contains values, where it should be `undefined`**

**Comment by Nick135** [comment]
Is there any idea to fix the error in MetricsQL

**Comment by Nick135** [comment]
Looks like the problem has something to do with empty values in the returned time series.
The data where added on change to the data base. So there are not always values for a series at a spezific time.

**Comment by Nick135** [comment]
My current solution is to run a separate query for each `name`.
It's not pretty and it's inflexible, but it works.

Hope there is a more elegant solution.

**Comment by zekker6** [comment]
> But I get 'missing match[] arg'
> What do I wrong to export the raw data?

Decoding URI from your example shows:
```
requestURI: /api/v1/export?'end=2022-06-06T19%3A29%3A07+00%3A00'&'match[]={__name__!%3D}'&'start=2022-06-06T19%3A25%3A48+00%3A00
```

Seems like shell encoded `'` as a value for URI. Could you try the following:
```
curl 'localhost:8888/api/v1/export?match[]=\{__name__!=""\}&start=2022-06-06T19:25:48&end=2022-06-06T19:29:07
```

**Comment by Nick135** [comment]
This solf the problem with data export.
However, the real problem of querying with Metricsql and `or` operator still remains.

local (Ubuntu):
```
curl localhost:8428/api/v1/export -d 'match[]={__name__!=""}' -d 'start=2023-01-14T06:01:28' -d 'end=2023-01-16T13:57:44'
```

WSL (Ubuntu):
```
curl http://<victoriametrics-addr>:8428/api/v1/export -d 'match[]={__name__!=""}' -d 'start=2023-01-14T06:01:28' -d 'end=2023-01-16T13:57:44'
```

PowerShell (Windows):
```
curl 'http://<victoriametrics-addr>:8428/api/v1/export?end=2023-01-16T13:57:44&match[]={__name__!=""}&start=2023-01-14T06:01:28'
```

**Comment by zekker6** [comment]
Could you share exported data so that we could check what causes this issue?

**Comment by Nick135** [comment]
'start=2023-01-07T08:27:30' 'end=2023-01-07T11:45:17'

[data.json.gz](https://github.com/VictoriaMetrics/VictoriaMetrics/files/10817121/data.json.gz)

**Comment by zekker6** [comment]
@Nick135 I've just ran a few tests to reproduce this issue and couldn't reproduce an issue in initially posted graph.
I've used data you've shared and ran the following query:
```
WITH (
  data = {__name__=~"SN003:1_[A-Z].*",unit="°C"} > 0 < 120,
  data1 = {__name__=~"SN003:1_[A-Z].*",unit="°C"} > 120 < 100
)
(data or (data1/10))
```
This resulted into the following graph:
![1678113029](https://user-images.githubusercontent.com/1367798/223139185-8ce65878-cce9-4fec-b343-91ff3d62c2f5.png)

Here are raw values for `{__name__=~"SN003:1_[A-Z].*",unit="°C"}` query:
![1678113088](https://user-images.githubusercontent.com/1367798/223139746-8d296a06-e725-4146-807a-14458d15a8c8.png)

So in this case `or` filled values which were missing in `data` with results of `data1` and it is what or is expected to do(you can give a read to [this](https://prometheus.io/docs/prometheus/latest/querying/operators/#logical-set-binary-operators) doc).

Could you elaborate on what do you think is behaving wrong in this case?

**Comment by Nick135** [comment]
Which tool do you use to plot the data?

With Grafana and Prometheus plugin I get this result:
![image](https://user-images.githubusercontent.com/14072733/223501642-55c05abc-fb60-4b71-8f85-85422e4e1b25.png)

Run separate query for each name:
![image](https://user-images.githubusercontent.com/14072733/223501981-9fa3fe9b-6233-4a26-8fad-745a06968e39.png)

**Comment by zekker6** [comment]
I've been using [VMUI](https://docs.victoriametrics.com/Single-server-VictoriaMetrics.html#vmui) bulit-in UI.
Just tried same query with Grafana and got the following:
![1678462419](https://user-images.githubusercontent.com/1367798/224357802-daf46326-7d8d-4639-a284-ed3592117554.png)
Don't see such resets here as well. Could you try to run query with [query tracing](https://docs.victoriametrics.com/Single-server-VictoriaMetrics.html#query-tracing) and share results?

**Comment by Nick135** [comment]
'query tracing' with vmui
![image](https://user-images.githubusercontent.com/14072733/224747096-419f69bb-64b9-4da7-b730-086456b0e22d.png)

```
{
  "duration_msec": 0.926,
  "message": "victoria-metrics-20230118-200413-tags-v1.86.2-0-g9785b3f57: /prometheus/api/v1/query_range: start=1673080200000, end=1673088300000, step=25000, query=\"WITH (\\n    data = {__name__=~\\\"SN003:1_[A-Z].*\\\",unit=\\\"°C\\\"} > 0 < 120,\\n    data1 = {__name__=~\\\"SN003:1_[A-Z].*\\\",unit=\\\"°C\\\"} > 120 < 1000\\n)\\n(data or (data1/10))\": series=5",
  "children": [
    {
      "duration_msec": 0.361,
      "message": "eval: query=(({__name__=~\"SN003:1_[A-Z].*\", unit=\"°C\"} > 0) < 120) or ((({__name__=~\"SN003:1_[A-Z].*\", unit=\"°C\"} > 120) < 1000) / 10), timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=5, points=1625, pointsPerSeries=325",
      "children": [
        {
          "duration_msec": 0.359,
          "message": "binary op \"or\": series=5",
          "children": [
            {
              "duration_msec": 0.312,
              "message": "execute left and right sides of \"or\" in parallel",
              "children": [
                {
                  "duration_msec": 0.275,
                  "message": "expr1",
                  "children": [
                    {
                      "duration_msec": 0.234,
                      "message": "eval: query=({__name__=~\"SN003:1_[A-Z].*\", unit=\"°C\"} > 0) < 120, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=5, points=1625, pointsPerSeries=325",
                      "children": [
                        {
                          "duration_msec": 0.232,
                          "message": "binary op \"<\": series=5",
                          "children": [
                            {
                              "duration_msec": 0.211,
                              "message": "eval: query={__name__=~\"SN003:1_[A-Z].*\", unit=\"°C\"} > 0, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=5, points=1625, pointsPerSeries=325",
                              "children": [
                                {
                                  "duration_msec": 0.209,
                                  "message": "binary op \">\": series=5",
                                  "children": [
                                    {
                                      "duration_msec": 0.179,
                                      "message": "eval: query={__name__=~\"SN003:1_[A-Z].*\", unit=\"°C\"}, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=5, points=1625, pointsPerSeries=325",
                                      "children": [
                                        {
                                          "duration_msec": 0.175,
                                          "message": "rollup default_rollup(): timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, window=0: neededMemoryBytes=0",
                                          "children": [
                                            {
                                              "duration_msec": 0.171,
                                              "message": "rollup cache get: query={__name__=~\"SN003:1_[A-Z].*\", unit=\"°C\"}, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, window=0",
                                              "children": [
                                                {
                                                  "duration_msec": 0.083,
                                                  "message": "load compressed entry from cache with size 4529 bytes"
                                                },
                                                {
                                                  "duration_msec": 0.08,
                                                  "message": "unpack the entry into 15882 bytes"
                                                },
                                                {
                                                  "duration_msec": 0.004,
                                                  "message": "unmarshal 5 series"
                                                },
                                                {
                                                  "duration_msec": 0.001,
                                                  "message": "return 5 series on a timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z]"
                                                }
                                              ]
                                            }
                                          ]
                                        }
                                      ]
                                    },
                                    {
                                      "duration_msec": 0.002,
                                      "message": "eval: query=0, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=1, points=325, pointsPerSeries=325"
                                    }
                                  ]
                                }
                              ]
                            },
                            {
                              "duration_msec": 0.001,
                              "message": "eval: query=120, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=1, points=325, pointsPerSeries=325"
                            }
                          ]
                        }
                      ]
                    }
                  ]
                },
                {
                  "duration_msec": 0.272,
                  "message": "expr2",
                  "children": [
                    {
                      "duration_msec": 0.253,
                      "message": "eval: query=(({__name__=~\"SN003:1_[A-Z].*\", unit=\"°C\"} > 120) < 1000) / 10, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=2, points=650, pointsPerSeries=325",
                      "children": [
                        {
                          "duration_msec": 0.252,
                          "message": "binary op \"/\": series=2",
                          "children": [
                            {
                              "duration_msec": 0.235,
                              "message": "eval: query=({__name__=~\"SN003:1_[A-Z].*\", unit=\"°C\"} > 120) < 1000, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=5, points=1625, pointsPerSeries=325",
                              "children": [
                                {
                                  "duration_msec": 0.233,
                                  "message": "binary op \"<\": series=5",
                                  "children": [
                                    {
                                      "duration_msec": 0.202,
                                      "message": "eval: query={__name__=~\"SN003:1_[A-Z].*\", unit=\"°C\"} > 120, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=5, points=1625, pointsPerSeries=325",
                                      "children": [
                                        {
                                          "duration_msec": 0.201,
                                          "message": "binary op \">\": series=5",
                                          "children": [
                                            {
                                              "duration_msec": 0.149,
                                              "message": "eval: query={__name__=~\"SN003:1_[A-Z].*\", unit=\"°C\"}, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=5, points=1625, pointsPerSeries=325",
                                              "children": [
                                                {
                                                  "duration_msec": 0.143,
                                                  "message": "rollup default_rollup(): timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, window=0: neededMemoryBytes=0",
                                                  "children": [
                                                    {
                                                      "duration_msec": 0.105,
                                                      "message": "rollup cache get: query={__name__=~\"SN003:1_[A-Z].*\", unit=\"°C\"}, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, window=0",
                                                      "children": [
                                                        {
                                                          "duration_msec": 0.016,
                                                          "message": "load compressed entry from cache with size 4529 bytes"
                                                        },
                                                        {
                                                          "duration_msec": 0.058,
                                                          "message": "unpack the entry into 15882 bytes"
                                                        },
                                                        {
                                                          "duration_msec": 0.027,
                                                          "message": "unmarshal 5 series"
                                                        },
                                                        {
                                                          "duration_msec": 0.002,
                                                          "message": "return 5 series on a timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z]"
                                                        }
                                                      ]
                                                    }
                                                  ]
                                                }
                                              ]
                                            },
                                            {
                                              "duration_msec": 0.013,
                                              "message": "eval: query=120, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=1, points=325, pointsPerSeries=325"
                                            }
                                          ]
                                        }
                                      ]
                                    },
                                    {
                                      "duration_msec": 0.001,
                                      "message": "eval: query=1000, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=1, points=325, pointsPerSeries=325"
                                    }
                                  ]
                                }
                              ]
                            },
                            {
                              "duration_msec": 0.004,
                              "message": "eval: query=10, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=1, points=325, pointsPerSeries=325"
                            }
                          ]
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    },
    {
      "duration_msec": 0.006,
      "message": "sort series by metric name and labels"
    },
    {
      "duration_msec": 0.496,
      "message": "generate /api/v1/query_range response for series=5, points=1547"
    }
  ]
}
```

**Comment by Nick135** [comment]
@zekker6: Do you find any problem in the query tracing?

**Comment by zekker6** [comment]
@Nick135 it seems like data you've shared in https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3710#issuecomment-1442267884 is missing some data points that are present in your DB.
There is the difference between query trace executed at your side and at fresh VM with imported data:
```
# From the trace above
eval: query=(({__name__=~\"SN003:1_[A-Z].*\", unit=\"°C\"} > 120) < 1000) / 10, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=2, points=650, pointsPerSeries=325

# From the local trace
eval: query=(({__name__=~"SN003:1_[A-Z].*", unit="°C"} > 120) < 100) / 10, timeRange=[2023-01-07T08:30:00Z..2023-01-07T10:45:00Z], step=25000, mayCache=true: series=0, points=0, pointsPerSeries=0
```
So as fas as I can see exported data sample returns `series=0, points=0` for this query. Could you check if all series are included into export above?

**Comment by Nick135** [comment]
`:8428/api/v1/export -d 'match[]={__name__!=""}' -d 'start=2023-01-07T08:27:30' -d 'end=2023-01-07T11:45:17'`

[data.json.gz](https://github.com/VictoriaMetrics/VictoriaMetrics/files/11101987/data.json.gz)

The old export was with filter on the series, which are used in the query.
**'match[]={__name__=~"SN003:1_[A-Z].*",unit="°C"}'**

**Comment by zekker6** [comment]
@Nick135 Thank you for sharing!
I was able to reproduce this issue now, will get back to you with more details after deeper investigation.

**Comment by Nick135** [comment]
@zekker6 Thanks a lot for your help!

My biggest misunderstanding was that ```or``` also takes values from vector2 that belong to another label and fill the labels in vector1 with those values. The ```union``` function does what I expected from the ```or``` operator. Vector 1 or Vector 2, but without mixing label values.

**Comment by valyala** [comment]
FYI, the commit e2367b6d1c5dcb941ad5dc24788ffbcbb81694de adds ability to preserve metric names after performing arithmetic operations in [MetricsQL](https://docs.victoriametrics.com/MetricsQL.html) - just add `keep_metric_names` after the operation. For example, the query from the original message can be rewritten now to the following one in order to get the expected graphs:

```metricsql
WITH (
    data = {__name__=~"MyMeasurement_[A-Z].*",unit="°C"} > 0 < 120,
    data1 = {__name__=~"MyMeasurement_[A-Z].*",unit="°C"} > 120 < 1000
)
(data or (data1/10 keep_metric_names))
```

See [these docs](https://docs.victoriametrics.com/MetricsQL.html#keep_metric_names) for details.

This functionality will be included in the next release. In the mean time it is possible to evaluate and test this functionality by building VictoriaMetrics from the commit mentioned above according to [these docs](https://docs.victoriametrics.com/#how-to-build-from-sources). VictoriaMetrics cluster can be built from the commit 781947a7e293417324ad482c6e505c5020f19fa6 according to [these docs](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#building-from-sources).

**Comment by valyala** [comment]
FYI, the ability to keep metric names from the left side of binary operation is available in VictoriaMetrics starting from [v1.92.0](https://docs.victoriametrics.com/CHANGELOG.html). See [these docs](https://docs.victoriametrics.com/#how-to-build-from-sources) for details.
